File: web-app/frontend/lib/api/co2cost.ts
/**
 * Server-only fetch for the per-building CO₂ cost-allocation (CO2KostAufG) report.
 *
 * Reads GET /compliance/co2-cost/{id} (B2B JWT). The landlord/tenant split of the
 * heating-fuel CO₂ price for one building the caller may see. Types mirror backend
 * app/schemas/co2_cost.py.
 */

export type Co2StairTier = {
  tier: number
  min_kg_m2: number
  max_kg_m2: number | null
  landlord_pct: number
  tenant_pct: number
}

export type Co2UnitAllocation = {
  unit_id: string
  area_m2: number | null
  area_share_pct: number | null
  co2_kg: number | null
  cost_eur: number | null
  landlord_cost_eur: number | null
  tenant_cost_eur: number | null
}

export type Co2CostAllocation = {
  fabric_building_id: string
  building_name: string
  building_type: string
  country_code: string | null
  is_residential: boolean
  reporting_year: number | null
  has_data: boolean
  floor_area_m2: number | null
  area_basis: string
  heating_co2_tonnes: number | null
  energy_content_kwh: number | null
  co2_intensity_kg_m2: number | null
  gas_emission_factor_kg_kwh: number
  co2_price_eur_t: number
  co2_price_eur_t_ets2: number
  ets2_year: number
  total_co2_cost_eur: number | null
  total_co2_cost_eur_ets2: number | null
  model: string
  tier: number | null
  landlord_pct: number | null
  tenant_pct: number | null
  landlord_cost_eur: number | null
  tenant_cost_eur: number | null
  landlord_cost_eur_ets2: number | null
  tenant_cost_eur_ets2: number | null
  stair: Co2StairTier[]
  note: string
  data_source: string
  units: Co2UnitAllocation[]
  unit_count: number | null
}

type FetchResult<T> = { ok: true; data: T } | { ok: false; error: string }

function getBackendUrl(): string | null {
  return process.env.BACKEND_URL || null
}

export async function fetchCo2CostAllocation(
  accessToken: string,
  fabricBuildingId: string
): Promise<FetchResult<Co2CostAllocation>> {
  const backendUrl = getBackendUrl()
  if (!backendUrl) return { ok: false, error: "Server misconfigured" }
  if (!accessToken) return { ok: false, error: "Missing access token" }

  try {
    const res = await fetch(
      `${backendUrl}/compliance/co2-cost/${encodeURIComponent(fabricBuildingId)}`,
      {
        method: "GET",
        headers: { Authorization: `Bearer ${accessToken}`, Accept: "application/json" },
        cache: "no-store",
      }
    )
    if (!res.ok) {
      const body = await res.text().catch(() => "<no body>")
      if (res.status === 503 && body.includes("fabric_unavailable")) {
        return { ok: false, error: "fabric_unavailable" }
      }
      console.error(`[fetchCo2CostAllocation] Backend ${res.status}: ${body.slice(0, 200)}`)
      return { ok: false, error: `Backend ${res.status}` }
    }
    return { ok: true, data: (await res.json()) as Co2CostAllocation }
  } catch (err) {
    console.error("[fetchCo2CostAllocation] Network error:", err)
    return { ok: false, error: "Network error" }
  }
}

File: web-app/backend/app/schemas/co2_cost.py
"""CO₂ cost-allocation (CO2KostAufG) report schema.

Per-building landlord/tenant split of the heating-fuel CO₂ price under the German
Kohlendioxidkostenaufteilungsgesetz (CO2KostAufG). Residential/mixed buildings use
the statutory 10-step model on kg CO₂ per m² of LIVING AREA (Wohnfläche); non-
residential buildings use the flat 50/50 split (the NWG step model is not yet in
force). Decision-support, not legal advice — see ``note`` / ``data_source`` and the
report disclaimer.
"""
from pydantic import BaseModel


class Co2StairTier(BaseModel):
    """One step of the statutory residential 10-step model (Anlage zu § 7 CO2KostAufG)."""

    tier: int
    min_kg_m2: float
    max_kg_m2: float | None  # None = open-ended top step (≥ 52)
    landlord_pct: int
    tenant_pct: int


class Co2UnitAllocation(BaseModel):
    """One dwelling's area-pro-rata share of the building heating-CO₂ cost (Est.).

    The split is by living-area share (Wohnfläche pro-rata), NOT per-unit metering;
    the same building-wide step landlord/tenant ratio applies to every unit.
    """

    unit_id: str
    area_m2: float | None
    area_share_pct: float | None
    co2_kg: float | None
    cost_eur: float | None
    landlord_cost_eur: float | None
    tenant_cost_eur: float | None


class Co2CostAllocationResponse(BaseModel):
    fabric_building_id: str
    building_name: str
    building_type: str
    country_code: str | None
    is_residential: bool
    reporting_year: int | None
    has_data: bool

    # Inputs (heating-fuel basis: only fuels that incur the CO₂ price)
    floor_area_m2: float | None  # area used for the intensity (see area_basis)
    area_basis: str  # "Wohnfläche (living area)" | "gross floor area"
    heating_co2_tonnes: float | None  # Scope-1 gas + oil/diesel (levied fuels)
    energy_content_kwh: float | None  # delivered heating energy (Energiegehalt, kWh)
    co2_intensity_kg_m2: float | None  # = heating CO₂ ÷ area — drives the step
    gas_emission_factor_kg_kwh: float  # BEHG natural-gas EF used for estimates

    # Prices: 2026 national nEHS corridor (€55–65/t), and the EU ETS2 onset (2028).
    co2_price_eur_t: float
    co2_price_eur_t_ets2: float
    ets2_year: int

    total_co2_cost_eur: float | None
    total_co2_cost_eur_ets2: float | None

    # Allocation result
    model: str  # "stair_10" (residential) | "flat_50_50" (non-residential)
    tier: int | None  # 1..10 for residential, else None
    landlord_pct: int | None
    tenant_pct: int | None
    landlord_cost_eur: float | None
    tenant_cost_eur: float | None
    landlord_cost_eur_ets2: float | None
    tenant_cost_eur_ets2: float | None

    # Reference + provenance
    # Per-unit (dwelling) breakdown — residential only, area pro-rata ESTIMATE
    # (Wohnfläche share, not per-unit metering). Empty for non-residential or
    # buildings without unit data.
    units: list[Co2UnitAllocation] = []
    unit_count: int | None = None

    stair: list[Co2StairTier]
    note: str
    data_source: str
